### PR TITLE
Neovim support

### PR DESCRIPTION
**Pull request by leemhenson**
Hi. I'm having a little hack around adding neovim support to this plugin. I'm _not_ a vim plugin developer, and I'm cargo-culting a bunch of stuff here, so I'd like to get some feedback before I continue onwards. 

1) Would you accept a PR that added neovim support throughout this plugin?
2) If yes, is this approach acceptable or can you point me towards a better pattern of supporting both job apis?
3) Any code improvements are welcomed! Vimscript is not my preferred language.... 😄

### CODE REVIEW

File: ftplugin/javascript/flood.vim [hunk 1]
```diff
@@ -13,7 +13,7 @@ if get(b:, 'loaded_flood')
 endif
 
 " version check
-if !has('channel') || !has('job')
+if !has('nvim') && (!has('channel') || !has('job'))
```
Comment (heavenshell) [inline ftplugin/javascript/flood.vim:16]: Prefer check Vim first.
`(!has('channel') || !has('job')) && !has('nvim')`

File: autoload/flood/check.vim [hunk 2]
```diff
@@ -45,35 +45,70 @@ endfunction
...
+  if flood#has_callback('check', 'after_run')
+    call g:flood_callbacks['check']['after_run']()
+  endif
+endfunction
+
+let g:chunks = {}
+
+function! s:neovim_job_handler(job_id, data, event) dict
+  if a:event == 'stdout'
+    if !has_key(g:chunks, a:job_id)
```
Comment (heavenshell) [inline autoload/flood/check.vim:79]: what is `g:chunks`?
do we need global scope variable?

File: autoload/flood/check.vim [hunk 3]
```diff
@@ -94,10 +129,20 @@ function! flood#check#run(...) abort
   endif
 
   let mode = a:0 > 0 ? a:1 : 'r'
-  let cmd = printf('%s --json', flood#flowbin())
-  let s:job = job_start(cmd, {
-        \ 'callback': {c, m -> s:check_callback(c, m, mode)},
-        \ })
+  let cmd = [flood#flowbin(), '--json']
```
Comment (heavenshell) [inline autoload/flood/check.vim:132]: Bit a tricky.
Move this into `if has('nvim')` block.
And Vim's block, use `let cmd = printf('%s --json', flood#flowbin())`

File: autoload/flood/check.vim [hunk 4]
```diff
@@ -94,10 +129,20 @@ function! flood#check#run(...) abort
...
+
+  if has('nvim')
```
Comment (heavenshell) [inline autoload/flood/check.vim:134]: I think we can create `job executer` function at flood.vim

```viml
function! flood#job_executer(cmd, args, callback_handlers)
  if has('nvim')
    " nvim job start
  else
    " vim job start
  endif
endif
```

all `flood/*.vim`'s job_start should call `flood#job_executer`.

What do you think?

**Review by heavenshell — COMMENTED** [review]
Thank you very much!

>Would you accept a PR that added neovim support throughout this plugin?

humm, This is a tough one.
Cause I can't support neovim(I don't want spent a time).
So, if bug or problem raised at neovim, user may send PR him/her self or neovim user(like you) may fix issue.
If this policy is ok, I'll include your PRs.

>If yes, is this approach acceptable or can you point me towards a better pattern of supporting both job apis?

Reviewed 😊 

I'm planing to create e2e tests(maybe I can push it this weekend).
I think you can use e2e tests for not break current behavoir.

**Comment by leemhenson** [comment]
Have you seen https://github.com/prabirshrestha/async.vim ? How do you feel about using that to abstract away the differences?

**Comment by heavenshell** [comment]
No. I don't want to depend on any libraries(except test things).
- Just download `vim-flood` and user can use it is best
- Bundle library is complicated to me
  - Use submodule?
  - plugin manager?
  - download directry?

I think Vim does not have defact bundler thing.
So, I want to keep my plugin clean.

Thx for suggestion.

**Comment by leemhenson** [comment]
Hi. Thanks for providing feedback here. I realised I could get everything I wanted out of vim-flow if I rejigged a few things in my config. Also, I don't want to spend any more time with Vimscript! 😅 

Good luck anyway!

**Pull request closed without merging**